### ISSUE TITLE
[Bug] ChainStreamPoller doesn't handle reorgs

### ISSUE THREAD

**Issue by popzxc** [alloy-rs/alloy#1005]
### Component

provider, pubsub

### Operating System

None

### Describe the bug

`ChainStreamPoller` doesn't seem to account for reorgs at all, e.g. [here](https://github.com/alloy-rs/alloy/blob/main/crates/provider/src/chain.rs#L72) it always expects the chain to be continuous.
And to fix that we probably need to compare block tip against `next_yield` [here](https://github.com/alloy-rs/alloy/blob/main/crates/provider/src/chain.rs#L56), and if `block_number < next_yield`, we should do some kind of search for the last non-diverging block (so probably we need to poll not only for block number, but rather for block header to use root hashes for comparison).

See [this comment](https://github.com/alloy-rs/alloy/pull/997#discussion_r1663579367) for more context.

**Comment by popzxc** [comment]
I think I know how to fix this issue and will try to do it this week.

**Comment by gakonst** [comment]
Can we close this?

**Comment by popzxc** [comment]
Not really; the linked PR only implemented the `Heartbeat` part of reorg handling, while this ticket is about `ChainStreamPoller`.
But with said PR merged, I'm unblocked to work on it, and hope to do it this week.

**Comment by yash-atreya** [comment]
I believe this was resolved by https://github.com/alloy-rs/alloy/pull/1321

Closing optimistically
